### PR TITLE
feat(iot-actions-alpha): open search action in IoT topic rule

### PR DESCRIPTION
**Pull request by GavinZZ**
The CDK has no support to writes data to an Amazon OpenSearch Service via AWS IoT Rule. Adding the action that writes data to an Amazon OpenSearch Service to AWS IoT topic rule.

OpenSearch action for IoT Topic Rule documentation is here: https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-properties-iot-topicrule-opensearchaction.html.

Closes https://github.com/aws/aws-cdk/issues/17702.

Tested by deploying my own stack and confirmed that publishing a message to IoT topic rule will successfully write data to open search service.

----

*By submitting this pull request, I confirm that my contribution is made under the terms of the Apache-2.0 license*

### CODE REVIEW

File: packages/@aws-cdk/aws-iot-actions-alpha/lib/open-search-action.ts [hunk 1]
```diff
@@ -0,0 +1,59 @@
...
+  readonly index: string;
+
+  /**
+   * The type of document you are storing.
+   */
+  readonly type: string;
+}
+
+/**
+ * The action to writes data to an Amazon OpenSearch Service domain.
```
Comment (paulhcsun) [inline packages/@aws-cdk/aws-iot-actions-alpha/lib/open-search-action.ts:28]:
```suggestion
 * The action to write data to an Amazon OpenSearch Service domain.
```

File: packages/@aws-cdk/aws-iot-actions-alpha/README.md [hunk 2]
```diff
@@ -355,20 +355,41 @@ to an HTTPS endpoint when it is triggered:
...
+      { key: 'key1', value: 'value1' },
+    ],
+    auth: { serviceName: 'serviceName', signingRegion: 'us-east-1' },
+  }),
+);
+```
+
+## Write Data to Open Search Service
+
+The code snippet below creates an AWS IoT Rule that write data
```
Comment (paulhcsun) [inline packages/@aws-cdk/aws-iot-actions-alpha/README.md:377]:
```suggestion
The code snippet below creates an AWS IoT Rule that writes data
```

File: packages/@aws-cdk/aws-iot-actions-alpha/test/opensearch/integ.open-search-action.ts [hunk 3]
```diff
@@ -0,0 +1,41 @@
...
+new IntegTest(app, 'iot-opensearch-action-integ-test', {
+  testCases: [stack],
+  cdkCommandOptions: {
+    deploy: {
+      args: {
+        rollback: true,
+      },
+    },
+  },
+});
```
Comment (paulhcsun) [inline packages/@aws-cdk/aws-iot-actions-alpha/test/opensearch/integ.open-search-action.ts:41]:
```suggestion
});

```

**Review by paulhcsun — CHANGES_REQUESTED** [review]
Looks good to me, just some minor grammar fixes.

File: packages/@aws-cdk/aws-iot-actions-alpha/lib/open-search-action.ts [hunk 4]
```diff
@@ -0,0 +1,59 @@
...
+   * The type of document you are storing.
+   */
+  readonly type: string;
+}
+
+/**
+ * The action to write data to an Amazon OpenSearch Service domain.
+ */
+export class OpenSearchAction implements iot.IAction {
+  constructor(private readonly domain: opensearch.Domain, private readonly props: OpenSearchActionProps) {
```
Comment (aaythapa) [inline packages/@aws-cdk/aws-iot-actions-alpha/lib/open-search-action.ts:31]: Probably my own lack of knowledge but why are the domain and props `private readonly`?

File: packages/@aws-cdk/aws-iot-actions-alpha/lib/open-search-action.ts [hunk 5]
```diff
@@ -0,0 +1,59 @@
...
+  /**
+   * The OpenSearch index where you want to store your data.
+   */
+  readonly index: string;
+
+  /**
+   * The type of document you are storing.
+   */
+  readonly type: string;
+}
```
Comment (aaythapa) [inline packages/@aws-cdk/aws-iot-actions-alpha/lib/open-search-action.ts:25]: why isn't `role` part of the interface?

File: packages/@aws-cdk/aws-iot-actions-alpha/lib/open-search-action.ts (hunk 4 above)
Comment (GavinZZ) [reply packages/@aws-cdk/aws-iot-actions-alpha/lib/open-search-action.ts:31]: that's the short format for defining the following
```
private const domain: opensearch.Domain;

constructor(domain: xx) {
  this.domain = domain;
}
```

File: packages/@aws-cdk/aws-iot-actions-alpha/lib/open-search-action.ts (hunk 5 above)
Comment (GavinZZ) [reply packages/@aws-cdk/aws-iot-actions-alpha/lib/open-search-action.ts:25]: it's part of `CommonActionProps`.

**Review by paulhcsun — APPROVED** [review]

**Pull request merged**